core/utils/codel.h

```diff
@@ -0,0 +1,301 @@
+#ifndef BESS_UTILS_CODEL_H_
+#define BESS_UTILS_CODEL_H_
+
+#include <cstring>
+#include <iostream>
+#include <string>
+
+#include "../kmod/llring.h"
+#include "../mem_alloc.h"
+#include "../packet.h"
+#include "time.h"
+
+#define DEFAULT_TARGET .05  // default delay target for codel
+#define DEFAULT_WINDOW .1   // default window size for codel
+#define DEFAULT_SLOTS 4096  // default number of slots in the codel queue
+
+/*
+  Codel(Controlled Delay Management) is an Queue controller based on this
+  article http://queue.acm.org/detail.cfm?id=2209336
+
+  It provides an active queue management to help prevent bufferbloat by dropping
+  queue entries at an increasing rate if the delay in the queue is above the
+  target queue delay. The equation used to calculate drop intervals is based on TCP
+  throughput response to drop probability.
+*/
+struct Codel {
+  double delay_target;  // the delay that codel will adjust for
+  double window;        // minimum time before changing state
+  // the time at which codel will change state to above target(0 if below)
+  double time_above_target;
+  double time_till_drop;  // the next time codel will drop
+  // the number of packets dropped while delay has been above target
+  uint32_t drop_count;
+  uint8_t dropping;  // whether in dropping state(above target for window)
+  llring* ring;      // queue
+  void (*drop_func)(void*);  // the function to call to drop a value
+};
```

Comment:
The algorithm cited in the paper does not use any floating point variables. Generally speaking forwarding code does not use floating point for performance reasons. If the "need" for floating point is to compute the square root of the count then this should be computable with an integer version of the square root function.